File: 백준/Silver/1699. 제곱수의 합/제곱수의 합.py
import sys

input = sys.stdin.readline

n = int(input())

dp = [i for i in range(n + 1)]  # 초기값을 i로 설정 (1로 다 더하는 경우)

# 제곱수 리스트 생성
squares = []
i = 1
while i * i <= n:
    squares.append(i * i)
    i += 1

# DP 테이블 갱신
for i in range(1, n + 1):
    for square in squares:
        if square > i:
            break
        dp[i] = min(dp[i], dp[i - square] + 1)

print(dp[n])

File: 백준/Silver/1699. 제곱수의 합/제곱수의 합.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        int[] dp = new int[n + 1]; 
        for (int i = 0; i <= n; i++) {
            dp[i] = i;
        }

        List<Integer> squares = new ArrayList<>();
        for (int i = 1; i * i <= n; i++) {
            squares.add(i * i);
        }

        for (int i = 2; i <= n; i++) {
            for (int square : squares) {
                if (square > i) break;
                dp[i] = Math.min(dp[i], dp[i - square] + 1);
            }
        }

        System.out.println(dp[n]);
    }
}
